### PR TITLE
HPCC-22467 assign ips for support node  even there is no any esp,roxie and thor req…

### PR DESCRIPTION
**Pull request by xwang2713**
…uested.

envgen doesn't assign ip for support if no esp, roxie and thor requested.
It is due to the original design for envgen which assume always create a desired environment.xml in one step. In complex environment like cloud it need several steps to generate a complete environment.xml. envgen2 is a tool to modify an existing environment.xml but it still use envgen libraries to create initial environment.xml.

Here is a sample usage to create an initial environment.xml for support nodes and dali without any esp, roxie and thor information:

envgen -env /tmp/environment.xml -ip 10.20.30.40 -supportnodes 1 -espnodes 0 -roxienodes 0 -thornodes 0 -assign_ips dali 10.20.30.10

Signed-off-by: Xiaoming Wang <xiaoming.wang@lexisnexi.com> 

@kenrowland please review

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:

Test build: http://10.240.32.243/view/custom7.x/job/CE-JIRA-HPCC-22467/
Tested function on local Ubuntu 18.04

### CODE REVIEW

File: deployment/deployutils/wizardInputs.cpp [hunk 1]
```diff
@@ -86,13 +86,26 @@ void CWizardInputs::setEnvironment()
   if(m_pXml->hasProp("@ipList"))
     formIPList(m_pXml->queryProp("@ipList"), m_ipaddress);
 
+  if(m_pXml->hasProp("@roxieNodes"))
+    m_roxieNodes = atoi(m_pXml->queryProp("@roxieNodes"));
+
```
Comment (kenrowland) [inline deployment/deployutils/wizardInputs.cpp]: I don't see where m_roxieNodes is initialized. I looked in the header, but perhaps I missed it. Same with m_thorNodes below.

File: deployment/deployutils/wizardInputs.cpp [hunk 2]
```diff
@@ -86,13 +86,26 @@ void CWizardInputs::setEnvironment()
...
   {
     m_supportNodes = atoi(m_pXml->queryProp("@supportNodes"));
 
     if (m_supportNodes)
     {
-      if (m_ipaddress.length() > 0 && m_ipaddress.length() > m_supportNodes)
+      if (m_ipaddress.length() > 0 &&
+         ((m_ipaddress.length() > m_supportNodes) ||
+         ((m_ipaddress.length() == m_supportNodes) &&
+          (m_roxieNodes == 0) && (m_thorNodes == 0) && (m_espNodes == 0))))
```
Comment (kenrowland) [inline deployment/deployutils/wizardInputs.cpp]: This is where I'd worry about m_roxieNodes and m_thorNodes not being initialized.

**Review by kenrowland — COMMENTED** [review]
@xwang2713 Please see comments

File: deployment/deployutils/wizardInputs.cpp (hunk 1 above)
Comment (xwang2713) [reply deployment/deployutils/wizardInputs.cpp]: envgen call deployutils.cpp function:
buildEnvFromWizard(optionsXml, pServiceName, pCfg, envXml, arrBuildSetWithAssignedIPs, arrAssignIPRanges, &dirMap);

optionalXml has supportNodes, roxieNodes, espNode and thorNodes variables
optionsXml.appendf("<XmlArgs supportNodes=\"%d\" roxieNodes=\"%d\" thorNodes=\"%d\" espNodes=\"%d\" slavesPerNode=\"%d\" thorChannelsPerSlave=\"%d\" roxieChannelsPerSlave=\"%d\" roxieOnDemand=\"%s\" ipList=\"%s\"/>", supportNodes, roxieNodes,
      thorNodes, espNodes, slavesPerNode, thorChannelsPerSlave, roxieChannelsPerSlave, roxieOnDemand?"true":"false", ipAddrs.str());

buildEnvFromWizard create CWizardInputs instance with all nodes numbers assigned to instance variable m_xxx

CWizardInputs wizardInputs(wizardXml, service, cfg, dirMap, arrBuildSetWithAssignedIPs, arrAssignedIPs);

Comment (kenrowland) [reply deployment/deployutils/wizardInputs.cpp]: @xwang2713 I will approve, but I still think you should initialize these values in the constructor initialization list.

**Review by kenrowland — APPROVED** [review]

**Comment by xwang2713** [comment]
@richardkchapman, this is ready to merge.

**Pull request merged**